include/thread_pool/thread_pool.hpp

```diff
@@ -75,64 +87,83 @@ class ThreadPoolImpl {
-    : m_workers(options.threadCount())
+inline ThreadPoolImpl<Task, Queue>::ThreadPoolImpl(ThreadPoolOptions options)
+    : m_idle_workers(options.threadCount())
+    , m_workers(options.threadCount())
+    , m_rouser(options.rousePeriod())
     , m_next_worker(0)
+    , m_num_busy_waiters(0)
 {
-    for(auto& worker_ptr : m_workers)
-    {
-        worker_ptr.reset(new Worker<Task, Queue>(options.queueSize()));
-    }
+    // Instatiate all workers.
+    for (auto it = m_workers.begin(); it != m_workers.end(); ++it)
+        it->reset(new Worker<Task, Queue>(options.busyWaitOptions(), options.queueSize()));
 
-    for(size_t i = 0; i < m_workers.size(); ++i)
-    {
-        m_workers[i]->start(i, &m_workers);
-    }
-}
+    // Initialize all worker threads.
+    for (size_t i = 0; i < m_workers.size(); ++i)
+        m_workers[i]->start(i, &m_workers, &m_idle_workers, &m_num_busy_waiters);
 
-template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>::ThreadPoolImpl(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
-{
-    *this = rhs;
+    m_rouser.start(&m_workers, &m_idle_workers, &m_num_busy_waiters);
 }
 
 template <typename Task, template<typename> class Queue>
 inline ThreadPoolImpl<Task, Queue>::~ThreadPoolImpl()
 {
     for (auto& worker_ptr : m_workers)
-    {
         worker_ptr->stop();
-    }
-}
 
-template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>&
-ThreadPoolImpl<Task, Queue>::operator=(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
-{
-    if (this != &rhs)
-    {
-        m_workers = std::move(rhs.m_workers);
-        m_next_worker = rhs.m_next_worker.load();
-    }
-    return *this;
+    m_rouser.stop();
 }
 
 template <typename Task, template<typename> class Queue>
 template <typename Handler>
 inline bool ThreadPoolImpl<Task, Queue>::tryPost(Handler&& handler)
 {
-    return getWorker().tryPost(std::forward<Handler>(handler));
+    // If there aren't busy waiters, let's see if we have any idling threads. 
+    // These incur higher overhead to wake up than the busy waiters.
+    if (m_num_busy_waiters.load(std::memory_order_acquire) == 0)
+    {
+        auto result = m_idle_workers.tryEmptyAny();
+        if (result.first)
+        {
+            auto success = m_workers[result.second]->tryPost(std::forward<Handler>(handler));
+            m_workers[result.second]->wake();
+
+            return success;
+        }
+    }
+
+    // No idle threads. Our threads are either active or busy waiting
+    // Either way, submit the work item in a round robin fashion.
+    if (!getWorker().tryPost(std::forward<Handler>(handler)))
+        return false; // Worker's task queue is full.
+
+    // We have to ensure that at least one thread is active after our submission.
+    // Threads could have transitioned into idling under our feet. We need to account for this.
+    if (m_num_busy_waiters.load(std::memory_order_acquire) == 0)
```

Comment:
The same issue is present here. Please carefully model cases when other thread changes the state of the variable.

Reply:
Same as above; these are soft checks that lower the likelihood of task dropping and improve the response time of posted tasks. The Rouser thread enforces the strict constraint on dropped tasks.

Reply:
I realize the comments above this check are misleading, will update.